### PR TITLE
feat: Github Workflow multi arch. Fixes #2080

### LINKED ISSUE #2080: Argo Workflow for ARM?
# Summary

Could the argo team create official images for ARM (armv7/arm64)?

There are various rogue docker images of say argo-cd for ARM by community members but nothing really official. Would be nice to truly support ARM.

# Motivation

ARM is extremely popular, even in the datacenter, and I would like to be able to use Argo in an multi-architecture cluster.

# Proposal

Change build infrastructure to build both amd64 and arm variants.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by dgiebert**
This PR introduces buildx into the CI/CD pipeline to allow building images for:
- linux/amd64
- linux/arm64
- linux/ppc64le
- linux/s390x

The pipeline is currently running on my fork, to see if everything works as expected.

It is based on this MR: https://github.com/argoproj/argo/pull/2491

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -17,6 +17,13 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
+      # https://github.com/marketplace/actions/docker-buildx
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: crazy-max/ghaction-docker-buildx@v3
```
Comment (alexec) [inline .github/workflows/docker-image.yml:23]: what is the licence for this? can you provide any information about security audits it has passed?

**Comment by alexec** [comment]
Can you please link to any issues this is related to or fixes?

File: .github/workflows/docker-image.yml (hunk 1 above)
Comment (dgiebert) [reply .github/workflows/docker-image.yml:23]: License is MIT: https://github.com/crazy-max/ghaction-docker-buildx/blob/master/LICENSE

I couldn't find anything regarding a security audit tbh

**Comment by dgiebert** [comment]
This fixes this prematurely closed Issue: https://github.com/argoproj/argo/issues/2080

As there obviously is no workflow to automatically generate ARM images

File: Dockerfile [hunk 2]
```diff
@@ -62,21 +63,24 @@ RUN apt-get update && \
...
 ####################################################################################################
 
-FROM node:14.0.0 as argo-ui
+FROM node:14.0.0-alpine as argo-ui
 
 ADD ["ui", "ui"]
 ADD ["api", "api"]
 
-RUN yarn --cwd ui install
+RUN apk add --no-cache git python2 make g++
```
Comment (alexec) [inline Dockerfile:82]: and this one

File: Dockerfile [hunk 3]
```diff
@@ -62,21 +63,24 @@ RUN apt-get update && \
...
 RUN ./recurl.sh /usr/local/bin/jq https://github.com/stedolan/jq/releases/download/jq-${JQ_VERSION}/jq-linux64
 RUN rm recurl.sh
+
 COPY hack/ssh_known_hosts /etc/ssh/ssh_known_hosts
 COPY --from=builder /usr/local/bin/docker /usr/local/bin/
 
 ####################################################################################################
 
-FROM node:14.0.0 as argo-ui
+FROM node:14.0.0-alpine as argo-ui
```
Comment (alexec) [inline Dockerfile:77]: can you add some commentary about this change please?

File: hack/image_arch.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+if [ "${IMAGE_OS}" = "linux" ]; then
```
Comment (alexec) [inline hack/image_arch.sh:2]: should add `set -eu -o pipefail` here

File: Dockerfile [hunk 5]
```diff
@@ -85,30 +89,30 @@ RUN yarn --cwd ui build
...
 
 # cli image
 RUN mkdir -p ui/dist
 COPY --from=argo-ui ui/dist/app ui/dist/app
 # stop make from trying to re-build this without yarn installed
 RUN touch ui/dist/node_modules.marker
 RUN touch ui/dist/app/index.html
-RUN make argo-server.crt argo-server.key dist/argo-linux-${IMAGE_ARCH}
-RUN ["sh", "-c", "./dist/argo-linux-${IMAGE_ARCH} version | grep clean"]
+RUN ./image_arch.sh && make argo-server.crt argo-server.key dist/argo-linux-${IMAGE_ARCH}
```
Comment (alexec) [inline Dockerfile:114]: I'm not sure this does what you want - environment variable set in a script don't get exported to the parent shell. Instead on line 88 you could try this:

```
RUN source ./image_arch.sh
```

File: .github/workflows/docker-image.yml (hunk 1 above)
Comment (alexec) [reply .github/workflows/docker-image.yml:23]: MIT is fine - thank you

File: Dockerfile (hunk 3 above)
Comment (dgiebert) [reply Dockerfile:77]: I can revert back the changes, that was just for me to reduce the time needed to download the image.
That also explains the timeout option for yarn ;)

Comment (alexec) [reply Dockerfile:77]: Is this faster? If so - please don't change it!

Comment (dgiebert) [reply Dockerfile:77]: I can test that, but I think if download speed is not an issue: No.

File: Dockerfile (hunk 5 above)
Comment (dgiebert) [reply Dockerfile:114]: Actually `source` does not exist: `#10 0.243 /bin/sh: 1: source: not found`

But you are right it does not export it outside the subshell

Comment (alexec) [reply Dockerfile:114]: try `.` instead of `source`?

Comment (dgiebert) [reply Dockerfile:114]: That works without pipefail, is that fine for you ?

Comment (alexec) [reply Dockerfile:114]: my bad `-o pipefail` only works in bash - so you should remove it

Comment (dgiebert) [reply Dockerfile:114]: Done, will push the changes

File: Dockerfile (hunk 2 above)
Comment (dgiebert) [reply Dockerfile]: reverted this one

File: Dockerfile (hunk 3 above)
Comment (dgiebert) [reply Dockerfile]: Also reverted this

**Comment by alexec** [comment]
Please `git merge master` to fix `codegen` issue and I will merge.

**Comment by alexec** [comment]
🚀

**Comment by dgiebert** [comment]
Seems there is an issue with pushing the images.
I will push a fix for that today !

**Pull request merged**